### ISSUE TITLE
Reproducing RAFT Results

### ISSUE THREAD

**Issue by m43** [deepmind/tapnet#17]
Thanks for the nice work! Is the codebase used to evaluate RAFT possibly available publicly?

I was using your [evaluate_dataset.py](https://github.com/deepmind/tapnet/blob/b8e0641e3c6a3483060e49df1def87fef16c8d1a/evaluation_datasets.py) utilities to load DAVIS and KUBRIC and to later compute the PCK metrics but wasn't able to reproduce the reported numbers. For example, my `average_pts_within_thresh` for DAVIS is `87.5%` and does not match the reported `46.3%`. The numbers that I've got for strided query evaluation with stride 5 are the following:

| RAFT Results              |     DAVIS |    KUBRIC |
|:--------------------------|----------:|----------:|
| ade_visible               |   1.42952 |  0.954813 |
| pts_within_0.01           |  20.585   | 20.6073   |
| pts_within_0.1            |  26.0005  | 31.6371   |
| pts_within_0.5            |  49.9805  | 68.3699   |
| pts_within_1              |  66.5964  | 82.676    |
| pts_within_2              |  82.5936  | 91.0452   |
| pts_within_4              |  92.3043  | 95.6189   |
| pts_within_8              |  96.7841  | 98.0127   |
| pts_within_16             |  99.0649  | 99.1844   |
| average_pts_within_thresh |  87.4687  | 93.3074   |

In case you might spot something obviously incorrect, here is the ground truth and my predicted trajectory for the first point and the first 5 frames of DAVIS, all of which were visible, alongside relevant summary metrics computed for that datapoint alone.
```js
*** Results dictionary for the first datapoint of RAFT:
iter: 1
video_idx: 0
point_idx_in_video: 0
trajectory_gt: tensor([[131.9137,  87.8248],
        [135.1333,  89.2444],
        [137.8000,  91.8519],
        [140.2000,  95.8815],
        [142.6000, 102.5185]])
trajectory_pred: tensor([[131.9137,  87.8248],
        [135.3244,  89.0901],
        [137.8152,  91.3863],
        [140.1928,  94.9903],
        [142.7608, 101.2055]])
visibility_gt: tensor([True, True, True, True, True])

*** Summary metrics for the first datapoint of RAFT:
idx: 1--0--0
ade_visible: 0.5850879549980164 (including query point)
pts_within_0.01: 0.0 (percentage)
pts_within_0.1: 0.0
pts_within_0.5: 50.0
pts_within_1: 75.0
pts_within_2: 100.0
pts_within_4: 100.0
pts_within_8: 100.0
pts_within_16: 100.0
```

Also, here are a few GIFs for a random data batch of DAVIS (top) and KUBRIC (bottom):

| Ground Truth | Prediction | Prediction on top of GT |
|:--:|:--:|:--:|
| ![1_5_None_raft_tapvid_davis_20 04 04___gt](https://user-images.githubusercontent.com/17498813/227731967-d9cbf0f6-865f-4b98-9cd1-14f7909c68f1.gif) | ![1_5_None_raft_tapvid_davis_20 04 04___pred](https://user-images.githubusercontent.com/17498813/227731976-197c4502-7c34-4c85-b21a-498e458cee15.gif) | ![1_5_None_raft_tapvid_davis_20 04 04___pred_on_gt](https://user-images.githubusercontent.com/17498813/227732114-8aab5bce-6034-42c3-97e6-fd71e9e7df4e.png) |
| ![1_5_None_raft_tapvid_rgb_stacking_22 21 57__gt](https://user-images.githubusercontent.com/17498813/227732418-2aa2886e-5dee-417f-9cbe-2dce2c845b5d.gif) | ![1_5_None_raft_tapvid_rgb_stacking_22 21 57__pred](https://user-images.githubusercontent.com/17498813/227732432-6adc8614-0e2d-43c6-b7b5-6c3777815461.gif) | ![1_5_None_raft_tapvid_rgb_stacking_22 21 57__pred_on_gt](https://user-images.githubusercontent.com/17498813/227732450-c28b8d4a-0b4c-4f91-adc9-e44c08115ae7.png) |

If relevant, these are the entry points into my codebase:
- [PyTorch DataLoader](https://github.com/m43/pips/blob/da6afe84117000ad8e1ba1788d0abc2e47048841/test_on_flt.py#L252-L253)
- [Dataset Wrapper](https://github.com/m43/pips/blob/da6afe84117000ad8e1ba1788d0abc2e47048841/datasets/tapviddataset.py#L47-L70)
- [Dataset Stride](https://github.com/m43/pips/blob/da6afe84117000ad8e1ba1788d0abc2e47048841/datasets/tapviddataset.py#L87-L110) (stride is called "sequence length")
- [RAFT loaded](https://github.com/m43/pips/blob/da6afe84117000ad8e1ba1788d0abc2e47048841/test_on_flt.py#L277-L279)
- [RAFT forward pass](https://github.com/m43/pips/blob/da6afe84117000ad8e1ba1788d0abc2e47048841/test_on_flt.py#L125-L146)
- [Metrics computation](https://github.com/m43/pips/blob/da6afe84117000ad8e1ba1788d0abc2e47048841/pips_utils/figures.py#L318-L334)
- Environment Setup: [part 1](https://github.com/m43/pips/blob/da6afe84117000ad8e1ba1788d0abc2e47048841/docs/reproduction.md), [part 2](https://github.com/m43/pips/blob/da6afe84117000ad8e1ba1788d0abc2e47048841/docs/evaluate_on_tapvid.md)

**Comment by serycjon** [comment]
I have also tried to reproduce RAFT results and although not exactly the same, the results were pretty close. The numbers you got are way too high. I think the 8-frame subsequences could be the reason? Didn't read your code carefully, but the evaluation should be: start from a query frame and track till the end of the sequence (not just few frames, even in the strided evaluation mode), then back in time till the frame 0.
Are you doing that? Could you show the visualization for the whole India DAVIS sequence?

**Comment by m43** [comment]
I was indeed not doing it till the end of the sequence, but just for subsequences of length 5 (and 8 when evaluating PIPs), which can easily explain the high accuracy. Will give it a shot!

**Comment by m43** [comment]
The latest figures show a smaller gap: DAVIS with strided queries evaluation now has `48.1%` compared to the reported `46.3%`. I am not sure if there are any missed details that could account for the `1.8%` disparity (e.g., how the mean is computed).

Summary table:

|         RAFT results                  |   DAVIS (first) |   DAVIS (strided) |
|:--------------------------|----------------:|------------------:|
| ade_visible               |      37.828     |        27.7633    |
| pts_within_0.01           |       0.0402818 |         0.0421326 |
| pts_within_0.1            |       1.83714   |         1.80536   |
| pts_within_0.5            |      11.1298    |        11.4926    |
| pts_within_1              |      17.4713    |        19.994     |
| pts_within_2              |      27.1705    |        32.2037    |
| pts_within_4              |      40.1074    |        48.2483    |
| pts_within_8              |      55.8723    |        64.5519    |
| pts_within_16             |      67.3868    |        75.4873    |
| average_pts_within_thresh |      41.6016    |        48.097     |

Visualization of a DAVIS video:

| Ground Truth | Prediction | Prediction on top of GT |
|:--:|:--:|:--:|
| ![individualImage](https://user-images.githubusercontent.com/17498813/227754675-d222f9b8-2ee2-41b6-a449-59757af9f0fb.gif) | ![individualImage](https://user-images.githubusercontent.com/17498813/227754682-bd9f5594-6dca-45da-b678-5b28ebd0ca1d.gif) | ![individualImage](https://user-images.githubusercontent.com/17498813/227754688-06daa4dd-848b-44e8-a8b0-b527e2f787f9.png) |

Updated [forward pass](https://github.com/m43/pips/blob/53f65ba601a36511ca5717bf4409981f8e096d8f/test_on_flt.py#L274-L329) and [dataset loading](https://github.com/m43/pips/blob/53f65ba601a36511ca5717bf4409981f8e096d8f/datasets/tapvid.py#L38-L73).

**Comment by m43** [comment]
Closing the issue since there is only a small disparity on the DAVIS subset. Thank you @serycjon for the help!
